cinema/views.py

```diff
@@ -48,6 +87,42 @@ class MovieSessionViewSet(viewsets.ModelViewSet):
     queryset = MovieSession.objects.all()
     serializer_class = MovieSessionSerializer
 
+    @staticmethod
+    def _param_to_date(query_string):
+        return datetime.datetime.strptime(query_string, "%Y-%m-%d")
+
+    def get_queryset(self):
+        queryset = self.queryset
+        if self.action == "list":
+            queryset = (
+                queryset
+                .select_related("movie", "cinema_hall")
+                .annotate(
+                    tickets_available=F("cinema_hall__rows")
+                    * F("cinema_hall__seats_in_row")
+                    - Count("tickets")
+                )
+            )
+        if self.action == "retrieve":
+            queryset = (
+                queryset
+                .select_related("movie", "cinema_hall")
+                .prefetch_related("tickets")
+            )
+
+        date = self.request.query_params.get("date")
+        if date:
+            queryset = queryset.filter(
+                show_time__date=date
```

Comment:
here we compare date obj with the str obj - this is usually not a good idea, however here it will work if the incoming data is of exactly right format